src/frontend/src/flows/register/captcha.ts

```diff
@@ -1,180 +1,225 @@
-    setTimeout(() => {
-      connection.createChallenge().then((cha) => {
-        resolve(cha);
-      });
-    });
-  });
-
-const init = (
-  connection: Connection,
-  identity: IdentifiableIdentity,
-  alias: string,
-  captcha: Promise<Challenge>
-): Promise<LoginFlowResult> => {
-  requestCaptcha(connection, captcha);
-
-  // since the index expects to regain control we unfortunately have to wrap
-  // this whole logic in a promise that then resolves (giving control back to
-  // the caller)
+export const promptCaptcha = ({
+  connection,
+  identity,
+  alias,
+  challenge,
+}: {
+  connection: Connection;
+  identity: IdentifiableIdentity;
+  alias: string;
+  challenge?: Promise<Challenge>;
+}): Promise<RegisterResult | LoginFlowCanceled> => {
   return new Promise((resolve) => {
-    const confirmRegisterButton = document.querySelector(
-      "#confirmRegisterButton"
-    ) as HTMLFormElement;
-    const captchaInput = document.querySelector(
-      "#captchaInput"
-    ) as HTMLFormElement;
-    const cancelButton = document.querySelector(
-      "#cancelButton"
-    ) as HTMLButtonElement;
-
-    cancelButton.onclick = () => {
-      resolve(cancel);
-    };
-
-    confirmRegisterButton.onclick = (e) => {
-      e.preventDefault();
-      e.stopPropagation();
-
-      const captchaStatusText = document.querySelector(
-        ".captcha-status-text"
-      ) as HTMLElement;
-      captchaStatusText.innerHTML = "Checking CAPTCHA challenge…";
-      confirmRegisterButton.disabled = true;
-
-      const captchaChars = captchaInput.value;
-      const captchaKey = confirmRegisterButton.dataset.captchaKey;
-
-      if (captchaKey === undefined) {
-        console.log("Something went wrong: no captcha key found");
-        requestCaptcha(connection);
-        return;
-      }
-
-      const challengeResult: ChallengeResult = {
-        key: captchaKey,
-        chars: captchaChars,
-      };
-
-      tryRegister(connection, identity, alias, challengeResult, resolve);
-    };
+    promptCaptchaPage({
+      cancel: () => resolve(cancel),
+      verifyChallengeChars: async ({ chars, challenge }) => {
+        const result = await connection.register(identity, alias, {
+          key: challenge.challenge_key,
+          chars,
+        });
+
+        switch (result.kind) {
+          case "badChallenge":
+            return badChallenge;
+          default:
+            return result;
+        }
+      },
+
+      requestChallenge: precomputedFirst(
+        // For the first call, use a pre-generated challenge
+        // if available.
+        challenge ?? connection.createChallenge(),
+        () => connection.createChallenge()
+      ),
+
+      onContinue: resolve,
+    });
   });
 };
+
+// Returns a function that returns `first` on the first call,
+// and values returned by `f()` from the second call on.
+function precomputedFirst<T>(first: T, f: () => T): () => T {
+  let firstTime = true;
+
+  return () => {
```

Comment:
nth: not sure, is the return type maybe missing here?

Reply:
Can you clarify the question? Missing from where?

Reply:
I meant should it be following (adding `: T`)?

```
  return (): T => {
```

not sure...